Fix stack traces for wasm inlined functions

justinmichaud opened the pull request:
#### d32c0d56c8d26375f1bcce2f1597bc0817651431
<pre>
Fix stack traces for wasm inlined functions
<a href="https://bugs.webkit.org/show_bug.cgi?id=253685">https://bugs.webkit.org/show_bug.cgi?id=253685</a>
rdar://106532509

Reviewed by Yusuke Suzuki.

Wasm inlining broke stack traces. We now store our call site index
at the beginning of an inlined function if we otherwise would have skipped
this. Like before, we continue storing the call site index at
places where we can throw if any function in our current stack frame has
exception handlers.

We add a map to wasm jit callees to go from CallSiteIndex to function name.

We separate out the code to read a wasm frame from the code to read JS
or native frames, and consult this map inside StackVisitor.

* JSTests/wasm/function-tests/stack-trace.js:
* JSTests/wasm/stress/simple-inline-stacktrace.js: Added.
(a.doThrow):
(i.catch):
* JSTests/wasm/stress/simple-inline-stacktrace.wasm: Added.
* JSTests/wasm/stress/simple-inline-stacktrace.wat: Added.
* Source/JavaScriptCore/interpreter/CallFrameInlines.h:
(JSC::CallFrame::codeBlock const):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::findExceptionHandler):
* Source/JavaScriptCore/interpreter/ShadowChicken.cpp:
(JSC::ShadowChicken::update):
* Source/JavaScriptCore/interpreter/StackVisitor.cpp:
(JSC::StackVisitor::gotoNextFrame):
(JSC::StackVisitor::unwindToMachineCodeBlockFrame):
(JSC::StackVisitor::readFrame):
(JSC::StackVisitor::readNonInlinedFrame):
(JSC::StackVisitor::readInlinedWasmFrame):
(JSC::StackVisitor::readInlinedFrame):
(JSC::StackVisitor::Frame::calleeSaveRegistersForUnwinding):
(JSC::StackVisitor::Frame::createArguments):
(JSC::StackVisitor::Frame::setToEnd):
(JSC::StackVisitor::Frame::dump const):
* Source/JavaScriptCore/interpreter/StackVisitor.h:
(JSC::StackVisitor::Frame::inlineCallFrame const):
(JSC::StackVisitor::Frame::isInlinedDFGFrame const):
(JSC::StackVisitor::Frame::isInlinedFrame const): Deleted.
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::B3IRGenerator):
(JSC::Wasm::B3IRGenerator::preparePatchpointForExceptions):
(JSC::Wasm::B3IRGenerator::createTailCallPatchpoint):
(JSC::Wasm::B3IRGenerator::emitInlineDirectCall):
(JSC::Wasm::B3IRGenerator::addCall):
(JSC::Wasm::parseAndCompileB3):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.h:
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::compileFunction):
* Source/JavaScriptCore/wasm/WasmBBQPlan.h:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::OptimizingJITCallee::addCodeOrigin):
(JSC::Wasm::OptimizingJITCallee::getOrigin const):
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmIRGeneratorHelpers.h:
(JSC::Wasm::PatchpointExceptionHandle::generate const):
* Source/JavaScriptCore/wasm/WasmNameSectionParser.cpp:
(JSC::Wasm::NameSectionParser::parse):

Canonical link: <a href="https://commits.webkit.org/261670@main">https://commits.webkit.org/261670@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/344092f3f0b8e34173204882a681e03642b45112

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112514 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21662 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1167 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4289 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/121078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/116580 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23000 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12823 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/5431 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118281 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/17080 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100286 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105568 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/99035 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/832 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/46087 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/100826 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14004 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/869 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/95134 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/12108 "Built successfully and passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/14691 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/10249 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/102292 "Built successfully") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/20030 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52874 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31957 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/8132 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16522 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/110331 "Built successfully") | 
| | | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/27257 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
<!--EWS-Status-Bubble-End-->

justinmichaud commented on JSTests/wasm/stress/simple-inline-stacktrace.wat:
```diff
@@ -0,0 +1,72 @@
+;; wat2wasm --enable-exceptions simple-inline-stacktrace.wat -o simple-inline-stacktrace.wasm --debug-names
```
Note: debug names are totally broken ATM, I might fix them in a subsequent patch.

Constellation commented on Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
```diff
@@ -4463,10 +4480,11 @@ auto B3IRGenerator::addCall(uint32_t functionIndex, const TypeDefinition& signat
     }
 
     if (callType == CallType::Call
+        && functionIndex - m_numImportFunctions != m_functionIndex
```
What is the purpose of this?

justinmichaud replied:
If we inline ourself recursively, then stack traces won't work using this method. At the same time though, inlining ourself recursively seems to me like it would rarely be helpful

Constellation replied:
OK.

Constellation commented on Source/JavaScriptCore/wasm/WasmCallee.cpp:
```diff
@@ -234,6 +235,33 @@ const WasmInstruction* LLIntCallee::outOfLineJumpTarget(const WasmInstruction* p
+    for (unsigned i = 0; i < codeOrigins.size(); ++i)
+        ASSERT(codeOrigins[i].lastInlineCSI <= lastInlineCSI);
+#endif
+    codeOrigins.append({ firstInlineCSI, lastInlineCSI, IndexOrName { functionIndex, info.nameSection->get(functionIndex) } });
```
Let's avoid adding `IndexOrName` since it is too costly in terms of size etc. `functionIndex` is enough.

Constellation commented on Source/JavaScriptCore/wasm/WasmCallee.h:
```diff
@@ -188,6 +198,7 @@ class OptimizingJITCallee : public JITCallee {
     void linkExceptionHandlers(Vector<UnlinkedHandlerInfo>, Vector<CodeLocationLabel<ExceptionHandlerPtrTag>>);
 
     StackMaps m_stackmaps;
+    Vector<WasmCodeOrigin, 0> codeOrigins = { };
```
`= { }` is not necessary.

Constellation commented on Source/JavaScriptCore/wasm/WasmCallee.cpp:
```diff
@@ -234,6 +235,37 @@ const WasmInstruction* LLIntCallee::outOfLineJumpTarget(const WasmInstruction* p
+{
+    isInlined = false;
+    for (auto& o : codeOrigins) {
+        if (o.firstInlineCSI <= csi && o.lastInlineCSI >= csi && !(depth--)) {
```
We can use binary search here first (lower_bounds)

Constellation reviewed (approved):
r=me

Commit:
Fix stack traces for wasm inlined functions
https://bugs.webkit.org/show_bug.cgi?id=253685
rdar://106532509

Reviewed by Yusuke Suzuki.

Wasm inlining broke stack traces. We now store our call site index
at the beginning of an inlined function if we otherwise would have skipped
this. Like before, we continue storing the call site index at
places where we can throw if any function in our current stack frame has
exception handlers.

We add a map to wasm jit callees to go from CallSiteIndex to function name.

We separate out the code to read a wasm frame from the code to read JS
or native frames, and consult this map inside StackVisitor.

* JSTests/wasm/function-tests/stack-trace.js:
* JSTests/wasm/stress/simple-inline-stacktrace.js: Added.
(a.doThrow):
(i.catch):
* JSTests/wasm/stress/simple-inline-stacktrace.wasm: Added.
* JSTests/wasm/stress/simple-inline-stacktrace.wat: Added.
* Source/JavaScriptCore/interpreter/CallFrameInlines.h:
(JSC::CallFrame::codeBlock const):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::findExceptionHandler):
* Source/JavaScriptCore/interpreter/ShadowChicken.cpp:
(JSC::ShadowChicken::update):
* Source/JavaScriptCore/interpreter/StackVisitor.cpp:
(JSC::StackVisitor::gotoNextFrame):
(JSC::StackVisitor::unwindToMachineCodeBlockFrame):
(JSC::StackVisitor::readFrame):
(JSC::StackVisitor::readNonInlinedFrame):
(JSC::StackVisitor::readInlinedWasmFrame):
(JSC::StackVisitor::readInlinedFrame):
(JSC::StackVisitor::Frame::calleeSaveRegistersForUnwinding):
(JSC::StackVisitor::Frame::createArguments):
(JSC::StackVisitor::Frame::setToEnd):
(JSC::StackVisitor::Frame::dump const):
* Source/JavaScriptCore/interpreter/StackVisitor.h:
(JSC::StackVisitor::Frame::inlineCallFrame const):
(JSC::StackVisitor::Frame::isInlinedDFGFrame const):
(JSC::StackVisitor::Frame::isInlinedFrame const): Deleted.
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.cpp:
(JSC::Wasm::B3IRGenerator::B3IRGenerator):
(JSC::Wasm::B3IRGenerator::preparePatchpointForExceptions):
(JSC::Wasm::B3IRGenerator::createTailCallPatchpoint):
(JSC::Wasm::B3IRGenerator::emitInlineDirectCall):
(JSC::Wasm::B3IRGenerator::addCall):
(JSC::Wasm::parseAndCompileB3):
* Source/JavaScriptCore/wasm/WasmB3IRGenerator.h:
* Source/JavaScriptCore/wasm/WasmBBQPlan.cpp:
(JSC::Wasm::BBQPlan::compileFunction):
* Source/JavaScriptCore/wasm/WasmBBQPlan.h:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::OptimizingJITCallee::addCodeOrigin):
(JSC::Wasm::OptimizingJITCallee::getOrigin const):
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmIRGeneratorHelpers.h:
(JSC::Wasm::PatchpointExceptionHandle::generate const):
* Source/JavaScriptCore/wasm/WasmNameSectionParser.cpp:
(JSC::Wasm::NameSectionParser::parse):

Canonical link: https://commits.webkit.org/261670@main
